File: python/invoices.py
"""
Processed-invoice tracker (for duplicate detection) + audit log.

Two files:
  invoices.json  — { "processed": [{invoice_number, vendor, message_id, ...}] }
  invoice_log.csv — append-only ledger of every action.
"""

import csv
import json
from datetime import datetime, timezone
from pathlib import Path

PROCESSED_FILE = Path("invoices.json")
LOG_FILE = Path("invoice_log.csv")

LOG_COLUMNS = [
    "timestamp_utc", "action", "vendor", "invoice_number",
    "amount", "currency", "due_date", "po_number", "status",
    "thread_id", "note",
]


def _load() -> dict:
    if not PROCESSED_FILE.exists():
        return {"processed": []}
    try:
        return json.loads(PROCESSED_FILE.read_text())
    except Exception:
        return {"processed": []}


def _save(state: dict) -> None:
    PROCESSED_FILE.write_text(json.dumps(state, indent=2))


def is_duplicate(invoice_number: str, vendor: str = "") -> bool:
    """Same invoice_number from the same vendor counts as duplicate.
    Without vendor, any matching invoice_number does."""
    if not invoice_number:
        return False
    state = _load()
    inv_clean = invoice_number.strip()
    v_clean = (vendor or "").strip().lower()
    for p in state.get("processed", []):
        if p.get("invoice_number", "").strip() == inv_clean:
            if not v_clean or (p.get("vendor", "").strip().lower() == v_clean):
                return True
    return False


def record_processed(invoice: dict) -> None:
    state = _load()
    state["processed"].append({
        "recorded_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        **invoice,
    })
    _save(state)


def _ensure_log_header() -> None:
    if not LOG_FILE.exists() or LOG_FILE.stat().st_size == 0:
        with LOG_FILE.open("w", newline="", encoding="utf-8") as f:
            csv.writer(f).writerow(LOG_COLUMNS)


def log_action(*, action: str, vendor: str = "", invoice_number: str = "",
               amount: str = "", currency: str = "", due_date: str = "",
               po_number: str = "", status: str = "", thread_id: str = "",
               note: str = "") -> None:
    _ensure_log_header()
    row = [
        datetime.now(timezone.utc).isoformat(timespec="seconds"),
        action, vendor, invoice_number, amount, currency, due_date,
        po_number, status, thread_id,
        note.replace("\n", " ")[:500],
    ]
    with LOG_FILE.open("a", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(row)

File: agentmail-invoice-processor/typescript/src/invoices.ts
/**
 * Processed-invoice tracker (for duplicate detection) + audit log.
 *
 * Two files:
 *   invoices.json  — { "processed": [{invoice_number, vendor, ...}] }
 *   invoice_log.csv — append-only ledger of every action.
 */

import * as fs from "node:fs";

const PROCESSED_FILE = "invoices.json";
const LOG_FILE = "invoice_log.csv";

const LOG_COLUMNS = [
  "timestamp_utc", "action", "vendor", "invoice_number",
  "amount", "currency", "due_date", "po_number", "status",
  "thread_id", "note",
] as const;

interface Processed {
  recorded_at: string;
  vendor: string;
  invoice_number: string;
  amount: number;
  currency: string;
  due_date: string;
  po_number: string;
  po_match: string;
  status: string;
  is_urgent: boolean;
  message_id: string;
  thread_id: string;
}

function loadProcessed(): { processed: Processed[] } {
  if (!fs.existsSync(PROCESSED_FILE)) return { processed: [] };
  try {
    return JSON.parse(fs.readFileSync(PROCESSED_FILE, "utf8"));
  } catch {
    return { processed: [] };
  }
}

function saveProcessed(state: { processed: Processed[] }): void {
  fs.writeFileSync(PROCESSED_FILE, JSON.stringify(state, null, 2));
}

export function isDuplicate(invoiceNumber: string, vendor: string = ""): boolean {
  if (!invoiceNumber) return false;
  const state = loadProcessed();
  const invClean = invoiceNumber.trim();
  const vClean = (vendor || "").trim().toLowerCase();
  for (const p of state.processed) {
    if ((p.invoice_number || "").trim() === invClean) {
      if (!vClean || (p.vendor || "").trim().toLowerCase() === vClean) {
        return true;
      }
    }
  }
  return false;
}

export function recordProcessed(invoice: Omit<Processed, "recorded_at">): void {
  const state = loadProcessed();
  state.processed.push({
    recorded_at: new Date().toISOString(),
    ...invoice,
  });
  saveProcessed(state);
}

function csvEscape(s: string): string {
  if (s == null) return "";
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function ensureLogHeader(): void {
  if (!fs.existsSync(LOG_FILE) || fs.statSync(LOG_FILE).size === 0) {
    fs.writeFileSync(LOG_FILE, LOG_COLUMNS.join(",") + "\n");
  }
}

export function logAction(opts: {
  action: string;
  vendor?: string;
  invoice_number?: string;
  amount?: string;
  currency?: string;
  due_date?: string;
  po_number?: string;
  status?: string;
  thread_id?: string;
  note?: string;
}): void {
  ensureLogHeader();
  const row = [
    new Date().toISOString(),
    opts.action,
    opts.vendor || "",
    opts.invoice_number || "",
    opts.amount || "",
    opts.currency || "",
    opts.due_date || "",
    opts.po_number || "",
    opts.status || "",
    opts.thread_id || "",
    (opts.note || "").replace(/\n/g, " ").slice(0, 500),
  ].map(csvEscape).join(",");
  fs.appendFileSync(LOG_FILE, row + "\n");
}
